src/Haskell/Ide/Engine/Plugin/ApplyRefact.hs

```diff
@@ -106,6 +122,103 @@ applyAllCmd uri = pluginGetFile "applyAll: " uri $ \fp -> do
 
 -- ---------------------------------------------------------------------
 
+newtype HlintDataDir = HlintDataDir
+  { getHlintDataDir :: Maybe FilePath
+  }
+  deriving (Show, Eq, Read, Ord)
+
+
+instance ExtensionClass HlintDataDir where
+  initialValue = HlintDataDir Nothing
+
+data InitialisationError
+  = NoDataFiles FilePath
+  | InvalidDataDir FilePath
+  | Other
+
+ppInitialisationError :: InitialisationError -> String
+ppInitialisationError (NoDataFiles fp) =
+  "We could not find the data-files at: \"" <> fp <> "\""
+ppInitialisationError (InvalidDataDir fp) =
+  "The location \""
+    <> fp
+    <> "\" does not exist. We obtained that location from the environment variable: "
+    <> hlintDataDirEnvironmentVariable
+ppInitialisationError Other = "We could not determine why it failed."
+
+data Tool = Stack | Cabal
+
+showTool :: Tool -> String
+showTool Stack = "stack"
+showTool Cabal = "cabal"
+
+hlintName :: String
+hlintName = "hlint-" ++ VERSION_hlint
+
+hlintDataDirDefaultLocation :: MonadIO m => m FilePath
+hlintDataDirDefaultLocation = liftIO $ getXdgDirectory XdgData ("hie" </> "hlint")
+
+hlintDataDirEnvironmentVariable :: String
+hlintDataDirEnvironmentVariable = "HIE_HLINT_DATADIR"
+
+initializeHlint :: IdeGhcM (Either InitialisationError FilePath)
+initializeHlint = do
+  explicitDataDirLocation <- liftIO $ lookupEnv hlintDataDirEnvironmentVariable
+  hieHlintDataDir <- hlintDataDirDefaultLocation
+  dataDirExists <- liftIO (doesFileExist hieHlintDataDir)
+  result <- case explicitDataDirLocation of
+    Just explicitDir -> do
+      liftIO (doesFileExist explicitDir)
+        >>= \case
+          True ->
+            return $ Right explicitDir
+          False ->
+            return $ Left $ InvalidDataDir explicitDir
+    Nothing ->
+      if dataDirExists
+        then return $ Right hieHlintDataDir
+        else return $ Left $ NoDataFiles hieHlintDataDir
+
+  case result of
+    Right dataDir -> put $ HlintDataDir $ Just dataDir
+    _ -> return ()
+
+  return result
+
+downloadHlintDatafiles :: IO ()
+downloadHlintDatafiles = do
+  defaultLocation <- hlintDataDirDefaultLocation
+  availTools      <- liftIO
+    $ filterM (fmap isJust . findExecutable . showTool) [Stack, Cabal]
+  case availTools of
+    []       -> return ()
+    tool : _ -> withSystemTempDirectory "hie-hlint-datafiles" $ \tmpDir -> do
+      let toolArgs t = case t of
+            Cabal -> ["get", hlintName]
+            Stack -> ["unpack", hlintName]
```

Comment:
What if the cabal is not updates to know about the hlint version?
What if the hlint version is not on stackage? (very unlikely, just feels like not good software)

Reply:
Mmm and what about download the data file directly from https://raw.githubusercontent.com/ndmitchell/hlint/master/data/hlint.yaml?
Are the rest of files inside data needed for the hie usage?
Hardcoding the url is not ideal, but it uses fewer moving parts

Reply:
Yes, that is also an option. Just making sure that we get the one as per the version of hlint we are using in the build.  Probably the best option.